x-pack/metricbeat/module/iis/application_pool/reader.go

```diff
@@ -62,87 +63,107 @@ var appPoolCounters = map[string]string{
+		// check if the pdhexpandcounterpath/pdhexpandwildcardpath functions have expanded the counter successfully.
+		if len(childQueries) == 0 || (len(childQueries) == 1 && strings.Contains(childQueries[0], "*")) {
+			// covering cases when PdhExpandWildCardPathW returns no counter paths or is unable to expand and the ignore_non_existent_counters flag is set
+			r.log.Debugw("No counter paths returned but PdhExpandWildCardPathW returned no errors", "initial query", value,
+				logp.Namespace("perfmon"), "expanded query", childQueries)
 			continue
 		}
-		for _, count := range counters {
-			if err = re.Query.AddCounter(count, "", "float", true); err != nil {
-				return errors.Wrapf(err, `failed to add counter (query="%v")`, count)
+		for _, v := range childQueries {
+			if err := r.query.AddCounter(v, "", "float", len(childQueries) > 1); err != nil {
+				return errors.Wrapf(err, `failed to add counter (query="%v")`, v)
 			}
-			newQueries = append(newQueries, count)
-			re.WorkerProcesses[count] = key
+			r.workerProcesses[v] = key
 		}
 	}
-	err = re.Query.RemoveUnusedCounters(newQueries)
+	err = r.query.RemoveUnusedCounters(newQueries)
 	if err != nil {
 		return errors.Wrap(err, "failed removing unused counter values")
 	}
 	return nil
 }
 
-// fetch executes collects the query data and maps the counter values to events.
-func (re *Reader) fetch(names []string) ([]mb.Event, error) {
-	// refresh performance counter list
-	// Some counters, such as rate counters, require two counter values in order to compute a displayable value. In this case we must call PdhCollectQueryData twice before calling PdhGetFormattedCounterValue.
-	// For more information, see Collecting Performance Data (https://docs.microsoft.com/en-us/windows/desktop/PerfCtrs/collecting-performance-data).
-	// A flag is set if the second call has been executed else refresh will fail (reader.executed)
-	if re.hasRun || len(re.Query.Counters) == 0 {
-		err := re.initCounters(names)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed retrieving counters")
-		}
-	}
-	// if the ignore_non_existent_counters flag is set and no valid counter paths are found the Read func will still execute, a check is done before
-	if len(re.Query.Counters) == 0 {
+// read executes a query and returns those values in an event.
+func (r *Reader) read() ([]mb.Event, error) {
+	if len(r.applicationPools) == 0 {
+		r.executed = true
 		return nil, nil
 	}
+
 	// Some counters, such as rate counters, require two counter values in order to compute a displayable value. In this case we must call PdhCollectQueryData twice before calling PdhGetFormattedCounterValue.
 	// For more information, see Collecting Performance Data (https://docs.microsoft.com/en-us/windows/desktop/PerfCtrs/collecting-performance-data).
-	if err := re.Query.CollectData(); err != nil {
+	if err := r.query.CollectData(); err != nil {
 		return nil, errors.Wrap(err, "failed querying counter values")
```

Comment:
is a `r.query.Close()` needed here too?

Reply:
no, I do not want to close the query at that point as I still want it to retry at the next fetch (most of time the error clear out)

Reply:
👍 thanks for clarifying